lib/Reconciler.lua

```diff
@@ -1,542 +1,361 @@
+						elseif newChildElement ~= oldChildElement then
+							DEBUG_print("\t\tScheduling reconcile of child", key)
 
-	if isPrimitiveElement(newElement) then
-		-- Roblox Instance change
+							scheduleTask(tree, taskReconcileNode({
+								node = childNode,
+								toElement = newChildElement,
+							}))
+						end
+					end
 
-		local oldRef = oldElement.props[Core.Ref]
-		local newRef = newElement.props[Core.Ref]
+					for key in pairs(oldValue) do
+						local newChildElement = newValue[key]
+						local childNode = node.children[key]
 
-		-- Change the ref in one pass before applying any changes.
-		-- Roact doesn't provide any guarantees with regards to the sequencing
-		-- between refs and other changes in the commit phase.
-		if newRef ~= oldRef then
-			applyRef(oldRef, nil)
-			applyRef(newRef, instanceHandle._rbx)
+						if newChildElement == nil then
+							DEBUG_print("\t\tScheduling unmount of child", key)
+							scheduleTask(tree, taskUnmountNode({
+								node = childNode,
+							}))
+						end
+					end
+				else
+					DEBUG_print("\tSetting", prop, newValue)
+					node.rbx[prop] = newValue
+				end
+			end
 		end
 
-		-- Update properties and children of the Roblox object.
-		Reconciler._reconcilePrimitiveProps(oldElement, newElement, instanceHandle._rbx)
-		Reconciler._reconcilePrimitiveChildren(instanceHandle, newElement)
-
-		instanceHandle._element = newElement
-
-		return instanceHandle
-	elseif isFunctionalElement(newElement) then
-		instanceHandle._element = newElement
-
-		local rendered = newElement.component(newElement.props)
-		local newChild
-
-		if instanceHandle._child then
-			-- Transition from tree to tree, even if 'rendered' is nil
-			newChild = Reconciler._reconcileInternal(instanceHandle._child, rendered)
-		elseif rendered then
-			-- Transition from nil to new tree
-			newChild = Reconciler._mountInternal(
-				rendered,
-				instanceHandle._parent,
-				instanceHandle._key,
-				instanceHandle._context
-			)
+		for prop in pairs(fromElement.props) do
+			if not visitedProps[prop] then
+				-- TODO: Use real mechanism for setting default values
+				node.rbx[prop] = nil
+			end
 		end
+	end
+end
 
-		instanceHandle._child = newChild
+local function processTreeTasksAsync(tree, timeBudget)
+	if #tree.tasks == 0 then
+		return
+	end
 
-		return instanceHandle
-	elseif isStatefulElement(newElement) then
-		instanceHandle._element = newElement
+	DEBUG_warn(("-"):rep(80))
+	DEBUG_print("Async frame:", #tree.tasks, "tasks in queue; marker at", tree.taskIndex)
 
-		-- Stateful elements can take care of themselves.
-		instanceHandle._instance:_update(newElement.props)
+	tree.tasksRunning = true
 
-		return instanceHandle
-	elseif isPortal(newElement) then
-		if instanceHandle._rbx ~= newElement.props.target then
-			local parent = instanceHandle._parent
-			local key = instanceHandle._key
-			local context = instanceHandle._context
+	local startTime = tick()
+	local i = tree.taskIndex
```

Comment:
You're right, when there are errors in the task implementation, things go very wrong, and this would help salvage that by skipping broken tasks at least.